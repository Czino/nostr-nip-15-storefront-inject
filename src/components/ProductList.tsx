import { NDKTag } from '@nostr-dev-kit/ndk'
import { useNostrHooks } from 'nostr-hooks'
import { useState } from 'react'
import { getTagsFromUrl } from '../helpers/getTagsFromUrl'
import { useCustomNdk } from '../hooks/useCustomNdk'
import { useGetProductEvents } from '../hooks/useGetProductEvents'
import { Error } from './Error'
import { ProductItem } from './ProductItem'

type Props = {
    pubkey?: string
    tags?: NDKTag[]
    imageProxy?: string
    showPrice?: boolean
}
export const ProductList = ({ pubkey = '', tags, imageProxy, showPrice }: Props) => {
    const [customNdk] = useCustomNdk()
    const [searchTags] = useState(tags || getTagsFromUrl(window.location.href))

    useNostrHooks(customNdk)
    const { productInfos, eose } = useGetProductEvents({ pubkey, tags: searchTags })
    const availableProducts = productInfos.filter((productInfo) => productInfo?.quantity >= 0)
    if (!pubkey) return <Error>Product List: Missing attribute data-pubkey</Error>

    return (
        <div>
            <div className="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-4">
                {availableProducts.map((productInfo) => (
                    <ProductItem
                        key={productInfo.id}
                        product={productInfo}
                        width={400}
                        showPrice={showPrice}
                        imageProxy={imageProxy}
                    />
                ))}
            </div>
            {!eose && <div className="text-center">Loading...</div>}
            {eose && availableProducts.length === 0 && <div className="text-center">No products found</div>}
        </div>
    )
}
