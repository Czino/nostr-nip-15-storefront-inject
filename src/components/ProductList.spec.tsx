import { render } from '@testing-library/react'
import { productEvent } from '../../test/data/eventData'
import { mockUseSubscribe } from '../setupTests'
import { ProductList } from './ProductList'

describe('ProductList', () => {
    it('renders error when no pubkey is passed', () => {
        mockUseSubscribe.mockReturnValue({ events: [], eose: false })
        const { asFragment } = render(<ProductList />)
        expect(asFragment()).toMatchSnapshot()
    })
    it('renders product list correctly', () => {
        mockUseSubscribe.mockReturnValue({ events: [productEvent], eose: true })

        const { asFragment } = render(<ProductList pubkey="pubkey" />)
        expect(asFragment()).toMatchSnapshot()
    })
})
