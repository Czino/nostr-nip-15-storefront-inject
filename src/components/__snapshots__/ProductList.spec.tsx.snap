// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList renders error when no pubkey is passed 1`] = `
<DocumentFragment>
  <div
    class="bg-red-700 text-white p-4 rounded text-center"
  >
    Product List: Missing attribute data-pubkey
  </div>
</DocumentFragment>
`;

exports[`ProductList renders product list correctly 1`] = `
<DocumentFragment>
  <div>
    <div
      class="grid grid-cols-1 gap-4 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4"
    >
      <a
        class="grid grid-cols-1 gap-1"
        href="/product/Cat%20Balaclava%20-%20Cactus/0f1a3984-7a72-4c6f-b1dc-bc170b6c4d8e"
      >
        <div
          class="w-full relative aspect-[2/3] overflow-hidden"
        >
          <div
            class="absolute inset-0 flex justify-center items-center -z-1"
          >
            <svg
              class="animate-spin h-7 w-7 text-light-1"
              fill="none"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                class="opacity-25"
                cx="12"
                cy="12"
                r="10"
                stroke="currentColor"
                stroke-width="4"
              />
              <path
                class="opacity-75"
                d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
                fill="currentColor"
              />
            </svg>
          </div>
          <img
            alt=""
            class="w-full h-full object-cover"
            src="https://f004.backblazeb2.com/file/plebeian-market/P_listing_BHQQ_media_3041b54d1a0b9de92aa91999214b8848bc6951fe7afb5f9b3ead59d69d03aaf7.jpeg"
            width="400"
          />
          <img
            alt=""
            class="object-cover w-full h-full absolute top-0 opacity-0 hover:opacity-100 transition-opacity duration-500"
            src="https://f004.backblazeb2.com/file/plebeian-market/P_listing_BHQQ_media_3041b54d1a0b9de92aa91999214b8848bc6951fe7afb5f9b3ead59d69d03aaf8.jpeg"
            width="400"
          />
        </div>
        <div
          class="px-2"
        >
          <h3
            class="text-xl font-bold"
          >
            Cat Balaclava - Cactus
          </h3>
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
